release/src/router/mtd-utils/mkfs.jffs2.1
.TH MKFS.JFFS2 1
.SH NAME
mkfs.jffs2 \- Create a JFFS2 file system image from directory
.SH SYNOPSIS
.B mkfs.jffs2
[
.B -p,--pad[=SIZE]
]
[
.B -r,-d,--root
.I directory
]
[
.B -s,--pagesize=SIZE
]
[
.B -e,--eraseblock=SIZE
]
[
.B -c,--cleanmarker=SIZE
]
[
.B -n,--no-cleanmarkers
]
[
.B -o,--output
.I image.jffs2
]
[
.B -l,--little-endian
]
[
.B -b,--big-endian
]
[
.B -D,--devtable=FILE
]
[
.B -f,--faketime
]
[
.B -q,--squash
]
[
.B -U,--squash-uids
]
[
.B -P,--squash-perms
]
[
.B --with-xattr
]
[
.B --with-selinux
]
[
.B --with-posix-acl
]
[
.B -m,--compression-mode=MODE
]
[
.B -x,--disable-compressor=NAME
]
[
.B -X,--enable-compressor=NAME
]
[
.B -y,--compressor-priority=PRIORITY:NAME
]
[
.B -L,--list-compressors
]
[
.B -t,--test-compression
]
[
.B -h,--help
]
[
.B -v,--verbose
]
[
.B -V,--version
]
[
.B -i,--incremental
.I image.jffs2
]

.SH DESCRIPTION
The program
.B mkfs.jffs2
creates a JFFS2 (Second Journalling Flash File System) file system
image and writes the resulting image to the file specified by the
.B -o
option or by default to the standard output, unless the standard
output is a terminal device in which case mkfs.jffs2 will abort.

The file system image is created using the files and directories
contained in the directory specified by the option
.B -r
or the present directory, if the
.B -r
option is not specified.

Each block of the files to be placed into the file system image
are compressed using one of the avaiable compressors depending
on the selected compression mode.

File systems are created with the same endianness as the host,
unless the
.B -b
or
.B -l
options are specified.  JFFS2 driver in the 2.4 Linux kernel only
supported images having the same endianness as the CPU. As of 2.5.48,
the kernel can be changed with a #define to accept images of the
non-native endianness. Full bi-endian support in the kernel is not
planned.

It is unlikely that JFFS2 images are useful except in conjuction
with the MTD (Memory Technology Device) drivers in the Linux
kernel, since the JFFS2 file system driver in the kernel requires
MTD devices.
.SH OPTIONS
Options that take SIZE arguments can be specified as either
decimal (e.g., 65536), octal (0200000), or hexidecimal (0x1000).
.TP
.B -p, --pad[=SIZE]
Pad output to SIZE bytes with 0xFF.  If SIZE is not specified,
the output is padded to the end of the final erase block.
.TP
.B -r, -d, --root=DIR
Build file system from directory DIR.  The default is the current
directory.
.TP
.B -s, --pagesize=SIZE
Use page size SIZE.  The default is 4 KiB.  This size is the
maximum size of a data node.  Set according to target system's memory
management page size (NOTE: this is NOT related to NAND page size).
.TP
.B -e, --eraseblock=SIZE
Use erase block size SIZE.  The default is 64 KiB.  If you use a erase
block size different than the erase block size of the target MTD
device, JFFS2 may not perform optimally. If the SIZE specified is
below 4096, the units are assumed to be KiB.
.TP
.B -c, --cleanmarker=SIZE
Write \'CLEANMARKER\' nodes with the size specified. It is not
normally appropriate to specify a size other than the default 12
bytes.
.TP
.B -n, --no-cleanmarkers
Do not write \'CLEANMARKER\' nodes to the beginning of each erase
block. This option can be useful for creating JFFS2 images for
use on NAND flash, and for creating images which are to be used
on a variety of hardware with differing eraseblock sizes.
.TP
.B -o, --output=FILE
Write JFFS2 image to file FILE.  Default is the standard output.
.TP
.B -l, --little-endian
Create a little-endian JFFS2 image.  Default is to make an image
with the same endianness as the host.
.TP
.B -b, --big-endian
Create a big-endian JFFS2 image.  Default is to make an image
with the same endianness as the host.
.TP
.B -D, --devtable=FILE
Use the named FILE as a device table file, for including devices and
changing permissions in the created image when the user does not have
appropriate permissions to create them on the file system used as
source.
.TP
.B -f, --faketime
Change all file timestamps to \'0\' for regression testing.
.TP
.B -q, --squash
Squash permissions and owners, making all files be owned by root and
removing write permission for \'group\' and \'other\'.
.TP
.B -U, --squash-uids
Squash owners making all files be owned by root.
.TP
.B -P, --squash-perms
Squash permissions, removing write permission for \'group\' and \'other\'.
.TP
.B --with-xattr
Enables xattr, stuff all xattr entries into jffs2 image file.
.TP
.B --with-selinux
Enables xattr, stuff only SELinux Labels into jffs2 image file.
.TP
.B --with-posix-acl
Enable xattr, stuff only POSIX ACL entries into jffs2 image file.
.TP
.B -m, --compression-mode=MODE
Set the default compression mode. The default mode is
.B priority
which tries the compressors in a predefinied order and chooses the first
successful one. The alternatives are:
.B none
(mkfs will not compress) and
.B size
(mkfs will try all compressor and chooses the one which have the smallest result).
.TP
.B -x, --disable-compressor=NAME
Disable a compressor. Use
.B -L
to see the list of the avaiable compressors and their default states.
.TP
.B -X, --enable-compressor=NAME
Enable a compressor. Use
.B -L
to see the list of the avaiable compressors and their default states.
.TP
.B -y, --compressor-priority=PRIORITY:NAME
Set the priority of a compressor. Use
.B -L
to see the list of the avaiable compressors and their default priority.
Priorities are used by priority compression mode.
.TP
.B -L, --list-compressors
Show the list of the avaiable compressors and their states.
.TP
.B -t, --test-compression
Call decompress after every compress - and compare the result with the original data -, and
some other check.
.TP
.B -h, --help
Display help text.
.TP
.B -v, --verbose
Verbose operation.
.TP
.B -V, --version
Display version information.
.TP
.B -i, --incremental=FILE
Generate an appendage image for FILE. If FILE is written to flash and flash
is appended with the output, then it seems as if it was one thing.

.SH LIMITATIONS
The format and grammar of the device table file does not allow it to
create symbolic links when the symbolic links are not already present
in the root working directory.

However, symbolic links may be specified in the device table file
using the \fIl\fR type for the purposes of setting their permissions
and ownership.
.SH BUGS
JFFS2 limits device major and minor numbers to 8 bits each.  Some
consider this a bug.

.B mkfs.jffs2
does not properly handle hard links in the input directory structure.
Currently, hard linked files will be expanded to multiple identical
files in the output image.
.SH AUTHORS
David Woodhouse
.br
Manual page written by David Schleef <ds@schleef.org>
.SH SEE ALSO
.BR mkfs (8),
.BR mkfs.jffs (1),
.BR fakeroot (1)


release/src/router/util-linux/disk-utils/mkfs.bfs.c
/*
 *  mkfs.bfs - Create SCO BFS filesystem - aeb, 1999-09-07
 *
 */

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <time.h>
#include <unistd.h>

#include "blkdev.h"
#include "c.h"
#include "nls.h"
#include "strutils.h"
#include "xalloc.h"

#define BFS_ROOT_INO		2
#define BFS_NAMELEN		14
#define BFS_BLOCKSIZE		512
#define BFS_SUPER_MAGIC		0x1badface

/* superblock - 512 bytes */
struct bfssb {
	unsigned int s_magic;
	unsigned int s_start;	/* byte offset of start of data */
	unsigned int s_end;	/* sizeof(slice)-1 */

	/* for recovery during compaction */
	int s_from, s_to;	/* src and dest block of current transfer */
	int s_backup_from, s_backup_to;

	/* labels - may well contain garbage */
	char s_fsname[6];
	char s_volume[6];
	char s_pad[472];
};

/* inode - 64 bytes */
struct bfsi {
	unsigned short i_ino;
	unsigned char i_pad1[2];
	unsigned long i_first_block;
	unsigned long i_last_block;
	unsigned long i_bytes_to_end;
	unsigned long i_type;	/* 1: file, 2: the unique dir */
	unsigned long i_mode;
	unsigned long i_uid, i_gid;
	unsigned long i_nlinks;
	unsigned long i_atime, i_mtime, i_ctime;
	unsigned char i_pad2[16];
};

#define BFS_DIR_TYPE	2

/* directory entry - 16 bytes */
struct bfsde {
	unsigned short d_ino;
	char d_name[BFS_NAMELEN];
};

static void __attribute__ ((__noreturn__)) usage(FILE * out)
{
	fprintf(out,
		_("Usage: %s [options] device [block-count]\n"),
		program_invocation_short_name);
	fprintf(out, _("\nOptions:\n"
		       " -N, --inodes=NUM    specify desired number of inodes\n"
		       " -V, --vname=NAME    specify volume name\n"
		       " -F, --fname=NAME    specify file system name\n"
		       " -v, --verbose       explain what is being done\n"
		       " -c                  this option is silently ignored\n"
		       " -l                  this option is silently ignored\n"
		       " -V, --version       output version information and exit\n"
		       "                     -V as version must be only option\n"
		       " -h, --help          display this help and exit\n\n"));

	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

static void __attribute__ ((__noreturn__)) print_version(void)
{
	printf(_("%s from %s\n"), program_invocation_short_name, PACKAGE_STRING);
	exit(EXIT_SUCCESS);
}

int main(int argc, char **argv)
{
	char *device, *volume, *fsname;
	long inodes;
	unsigned long long total_blocks, ino_bytes, ino_blocks, data_blocks;
	unsigned long long user_specified_total_blocks = 0;
	int verbose = 0;
	int fd;
	struct bfssb sb;
	struct bfsi ri;
	struct bfsde de;
	struct stat statbuf;
	time_t now;
	int c, i, len;

	enum { VERSION_OPTION = CHAR_MAX + 1 };
	static const struct option longopts[] = {
		{"inodes", required_argument, NULL, 'N'},
		{"vname", required_argument, NULL, 'V'},
		{"fname", required_argument, NULL, 'F'},
		{"verbose", no_argument, NULL, 'v'},
		{"version", no_argument, NULL, VERSION_OPTION},
		{"help", no_argument, NULL, 'h'},
		{NULL, 0, NULL, 0}
	};

	if (argc < 2)
		usage(stderr);

	if (argc == 2 && !strcmp(argv[1], "-V"))
		print_version();

	volume = fsname = "      ";	/* is there a default? */
	inodes = 0;

	while ((c = getopt_long(argc, argv, "N:V:F:vhcl", longopts, NULL)) != -1) {
		switch (c) {
		case 'N':
			inodes = strtol_or_err(optarg, _("invalid number of inodes"));
			break;

		case 'V':
			len = strlen(optarg);
			if (len <= 0 || len > 6)
				errx(EXIT_FAILURE, _("volume name too long"));
			volume = xstrdup(optarg);
			break;

		case 'F':
			len = strlen(optarg);
			if (len <= 0 || len > 6)
				errx(EXIT_FAILURE, _("fsname name too long"));
			fsname = xstrdup(optarg);
			break;

		case 'v':
			verbose = 1;
			break;

		case 'c':
		case 'l':
			/* when called via mkfs we may get options c,l,v */
			break;

		case VERSION_OPTION:
			print_version();
		case 'h':
			usage(stdout);
		default:
			usage(stderr);
		}
	}

	if (optind == argc)
		usage(stderr);

	device = argv[optind++];

	if (stat(device, &statbuf) < 0)
		err(EXIT_FAILURE, _("cannot stat device %s"), device);

	if (!S_ISBLK(statbuf.st_mode))
		errx(EXIT_FAILURE, _("%s is not a block special device"), device);

	fd = open(device, O_RDWR | O_EXCL);
	if (fd < 0)
		err(EXIT_FAILURE, _("cannot open %s"), device);

	if (optind == argc - 1)
		user_specified_total_blocks =
			strtoll_or_err(argv[optind], _("invalid block-count"));
	else if (optind != argc)
		usage(stderr);

	if (blkdev_get_sectors(fd, &total_blocks) == -1) {
		if (!user_specified_total_blocks)
			err(EXIT_FAILURE, _("cannot get size of %s"), device);
		total_blocks = user_specified_total_blocks;
	} else if (user_specified_total_blocks) {
		if (user_specified_total_blocks > total_blocks)
			errx(EXIT_FAILURE,
			     _("blocks argument too large, max is %llu"),
			     total_blocks);
		total_blocks = user_specified_total_blocks;
	}

	if (!inodes) {
		/* pick some reasonable default */
		inodes = 8 * (total_blocks / 800);
		if (inodes < 48)
			inodes = 48;
		if (512 < inodes)
			inodes = 512;
	} else {
		/* believe the user */
		if (512 < inodes)
			errx(EXIT_FAILURE, _("too many inodes - max is 512"));
	}

	ino_bytes = inodes * sizeof(struct bfsi);
	ino_blocks = (ino_bytes + BFS_BLOCKSIZE - 1) / BFS_BLOCKSIZE;
	data_blocks = total_blocks - ino_blocks - 1;

	/* mimic the behaviour of SCO's mkfs - maybe this limit is needed */
	if (data_blocks < 32)
		errx(EXIT_FAILURE,
		     _("not enough space, need at least %llu blocks"),
		     ino_blocks + 33);

	memset(&sb, 0, sizeof(sb));
	sb.s_magic = BFS_SUPER_MAGIC;
	sb.s_start = ino_bytes + sizeof(struct bfssb);
	sb.s_end = total_blocks * BFS_BLOCKSIZE - 1;
	sb.s_from = sb.s_to = sb.s_backup_from = sb.s_backup_to = -1;
	memcpy(sb.s_fsname, fsname, 6);
	memcpy(sb.s_volume, volume, 6);

	if (verbose) {
		fprintf(stderr, _("Device: %s\n"), device);
		fprintf(stderr, _("Volume: <%-6s>\n"), volume);
		fprintf(stderr, _("FSname: <%-6s>\n"), fsname);
		fprintf(stderr, _("BlockSize: %d\n"), BFS_BLOCKSIZE);
		if (ino_blocks == 1)
			fprintf(stderr, _("Inodes: %lu (in 1 block)\n"),
				inodes);
		else
			fprintf(stderr, _("Inodes: %lu (in %llu blocks)\n"),
				inodes, ino_blocks);
		fprintf(stderr, _("Blocks: %lld\n"), total_blocks);
		fprintf(stderr, _("Inode end: %d, Data end: %d\n"),
			sb.s_start - 1, sb.s_end);
	}

	if (write(fd, &sb, sizeof(sb)) != sizeof(sb))
		err(EXIT_FAILURE, _("error writing superblock"));

	memset(&ri, 0, sizeof(ri));
	ri.i_ino = BFS_ROOT_INO;
	ri.i_first_block = 1 + ino_blocks;
	ri.i_last_block = ri.i_first_block +
	    (inodes * sizeof(de) - 1) / BFS_BLOCKSIZE;
	ri.i_bytes_to_end = ri.i_first_block * BFS_BLOCKSIZE
	    + 2 * sizeof(struct bfsde) - 1;
	ri.i_type = BFS_DIR_TYPE;
	ri.i_mode = S_IFDIR | 0755;	/* or just 0755 */
	ri.i_uid = 0;
	ri.i_gid = 1;			/* random */
	ri.i_nlinks = 2;
	time(&now);
	ri.i_atime = now;
	ri.i_mtime = now;
	ri.i_ctime = now;

	if (write(fd, &ri, sizeof(ri)) != sizeof(ri))
		err(EXIT_FAILURE, _("error writing root inode"));

	memset(&ri, 0, sizeof(ri));
	for (i = 1; i < inodes; i++)
		if (write(fd, &ri, sizeof(ri)) != sizeof(ri))
			err(EXIT_FAILURE, _("error writing inode"));

	if (lseek(fd, (1 + ino_blocks) * BFS_BLOCKSIZE, SEEK_SET) == -1)
		err(EXIT_FAILURE, _("seek error"));

	memset(&de, 0, sizeof(de));
	de.d_ino = BFS_ROOT_INO;
	memcpy(de.d_name, ".", 1);
	if (write(fd, &de, sizeof(de)) != sizeof(de))
		err(EXIT_FAILURE, _("error writing . entry"));

	memcpy(de.d_name, "..", 2);
	if (write(fd, &de, sizeof(de)) != sizeof(de))
		err(EXIT_FAILURE, _("error writing .. entry"));

	if (close(fd) < 0)
		err(EXIT_FAILURE, _("error closing %s"), device);

	return EXIT_SUCCESS;
}
